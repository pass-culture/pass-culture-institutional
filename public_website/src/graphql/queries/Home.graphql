query Home(
  $filters: ResourceFiltersInput
  $pagination: PaginationArg
  $sort: [String]
) {
  home {
    aboutSection {
      ...ComponentBlockCenteredText
    }
    CTASection {
      ...ComponentBlockPushCta
    }
    socialMediaSection {
      ...ComponentBlockSocialMedia
    }
    latestStudies {
      ...ComponentBlockLatestNews
    }
    eligibilitySection {
      ...ComponentHomeEligibilitySection
    }
    heroSection {
      ...ComponentHomeHeroSection
    }
    recommendationsSection {
      ...ComponentHomeRecommendationsSection
    }
    seo {
      ...ComponentSharedSeo
    }
  }
  resources(filters: $filters, pagination: $pagination, sort: $sort) {
    ...Resource
  }
}
