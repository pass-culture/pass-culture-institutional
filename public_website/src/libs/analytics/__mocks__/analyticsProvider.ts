import { vi } from 'vitest'

export const analyticsProvider = {
  init: vi.fn(),
  logEvent: vi.fn(),
}
