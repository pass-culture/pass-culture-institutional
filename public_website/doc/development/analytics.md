# Analytics

We use Firebase Analytics to track usage of the app. Please get in touch with an admin to get Firebase credentials.

## Usage

To track an new event, first add it to the `EventMap` in [analyticsProvider.ts](../../src/libs/analytics/analyticsProvider.ts):

```ts
export type EventMap = {
  // ... existing events
  my_event: my_params
}
```

Then, log it with the `logEvent` function:

```ts
import { analyticsProvider } from '@/libs/analytics/analyticsProvider'

analyticsProvider.logEvent('my_event', params)
```

## Manual testing

To test that the event is logged correctly, you can use the [Google Analytics Debugger](https://chrome.google.com/webstore/detail/google-analytics-debugger/jnkmfdileelhofjcijamephohjechhna) extension for Chrome. It will show you the events that are logged in the console.

With this extension, you can also view events logged in the DebugView in the Firebase console.

For further information, please check <https://firebase.google.com/docs/analytics/debugview#enable_debug_mode>.

## Unit testing

Cf. [Unit testing](./unit-tests/analytics.md)
