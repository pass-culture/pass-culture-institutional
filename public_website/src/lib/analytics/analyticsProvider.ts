import { Analytics, getAnalytics, logEvent } from 'firebase/analytics'
import { initializeApp } from 'firebase/app'

import { analyticsConfig } from '@/lib/analytics/config'

let analyticsInstance: Analytics

export type EventMap = {
  testEvent: { origin: 'testOrigin' }
  goToSignUpNative: {
    origin:
      | 'header'
      | 'home'
      | 'menu-young-people-and-parents'
      | 'get-your-credit'
      | 'simulator'
  }
  goToSignUpPro: {
    origin: 'header' | 'menu-pros' | 'essential-pros' | 'how-to-propose-offers'
  }
  goToLoginNative: { origin: 'header' }
  goToLoginPro: { origin: 'header' }
  downloadApp: { origin: 'home' | 'menu-young-people-and-parents' | 'footer' }
  goToFaqNative: {
    origin: 'get-your-credit' | 'help-young-people-and-parents' | 'parents'
  }
  goToFaqPro: { origin: 'help-pros' }
  contactSupport: { origin: 'help-pros' | 'help-teachers' }
  pageView: { origin: string }
}

export const eventMapKeys: { [K in keyof EventMap]: true } = {
  testEvent: true,
  goToSignUpNative: true,
  goToSignUpPro: true,
  goToLoginNative: true,
  goToLoginPro: true,
  downloadApp: true,
  goToFaqNative: true,
  goToFaqPro: true,
  contactSupport: true,
  pageView: true,
}

export enum eventOriginsEnum {
  header = 'header',
  home = 'home',
  'get-your-credit' = 'get-your-credit',
  simulator = 'simulator',
  'menu-pros' = 'menu-pros',
  'essential-pros' = 'essential-pros',
  'how-to-propose-offers' = 'how-to-propose-offers',
  'menu-young-people-and-parents' = 'menu-young-people-and-parents',
  footer = 'footer',
  'help-young-people-and-parents' = 'help-young-people-and-parents',
  parents = 'parents',
  'help-pros' = 'help-pros',
  'help-teachers' = 'help-teachers',
}

export const analyticsProvider = {
  init: () => {
    const app = initializeApp(analyticsConfig)
    analyticsInstance = getAnalytics(app)
  },
  logEvent: <K extends keyof EventMap>(eventName: K, options: EventMap[K]) =>
    logEvent<string>(analyticsInstance, eventName, options),
}
