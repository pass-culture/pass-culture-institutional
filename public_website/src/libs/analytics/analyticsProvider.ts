import { Analytics, getAnalytics, logEvent } from 'firebase/analytics'
import { initializeApp } from 'firebase/app'

import { analyticsConfig } from '@/libs/analytics/config'

let analyticsInstance: Analytics

type EventMap = {
  testEvent: { param: string }
  // ... add other event types here ...
}

export const analyticsProvider = {
  init: () => {
    const app = initializeApp(analyticsConfig)
    analyticsInstance = getAnalytics(app)
  },
  logEvent: <K extends keyof EventMap>(eventName: K, options: EventMap[K]) =>
    logEvent<string>(analyticsInstance, eventName, options),
}
