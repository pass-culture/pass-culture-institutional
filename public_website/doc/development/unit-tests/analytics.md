# Unit testing analytics

In the component you log the analytics event, you can mock the `logEvent` function:

```ts
import { analyticsProvider } from '@/libs/analytics/analyticsProvider'

vi.mock('@/libs/analytics/analyticsProvider')
const mockLogEvent = analyticsProvider.logEvent
```

Then, you can test that the event is logged correctly:

```ts
expect(mockLogEvent).toHaveBeenCalledWith('my_event', params)
```
