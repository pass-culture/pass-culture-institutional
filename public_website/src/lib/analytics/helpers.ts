import { analyticsProvider, EventMap, eventMapKeys } from './analyticsProvider'

function isKeyOfEventMap(key: string): key is keyof EventMap {
  return key in eventMapKeys
}

export const onClickAnalytics = (cta: {
  eventName?: string
  eventOrigin?: string
}) => {
  if (cta.eventName && cta.eventOrigin && isKeyOfEventMap(cta.eventName)) {
    analyticsProvider.logEvent(cta.eventName, {
      origin: cta.eventOrigin,
    })
  } else {
    // When Sentry is added to project: If the eventName is not in EventMap, do not send to Firebase and log to Sentry
    return undefined
  }
}
