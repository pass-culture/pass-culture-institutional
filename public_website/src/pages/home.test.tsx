import React from 'react'
import { describe, expect, it, vi } from 'vitest'
import { axe } from 'vitest-axe'

import Home from './index'
import { analyticsProvider } from '@/libs/analytics/analyticsProvider'
import { act, fireEvent, render, screen } from '@/tests'

vi.mock('@/libs/analytics/analyticsProvider')
const mockLogEvent = analyticsProvider.logEvent

describe('Home page', () => {
  it('should pass axe accessibility tests', async () => {
    const { container } = render(<Home />)

    let a11yResult
    await act(async () => {
      a11yResult = await axe(container)
    })
    expect(a11yResult).toHaveNoViolations()
  })

  it('should render the page', () => {
    const { container } = render(<Home />)

    expect(container).toMatchSnapshot()
  })

  it('should select the checkbox when clicking on its label', () => {
    render(<Home />)

    const checkbox = screen.getByTestId('checkbox-acceptTerms')
    const label = screen.getByText('Checkbox Ã  cocher')

    fireEvent.click(label)

    expect(checkbox).toHaveProperty('checked', true)
  })

  it('should trigger test event when clicking on the button', () => {
    render(<Home />)

    const button = screen.getByRole('button')
    fireEvent.click(button)

    expect(mockLogEvent).toHaveBeenCalledWith('testEvent', {
      param: 'testParam',
    })
  })
})
