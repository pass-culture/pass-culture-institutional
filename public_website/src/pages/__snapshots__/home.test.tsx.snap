// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page > should render the page 1`] = `
<div>
  <div
    class="sc-gtsqUy qFqbA"
  >
    <main
      class="sc-bdnyFh cOrhoO"
    >
      <h1
        class="sc-pNWRh djejpM"
      >
        Welcome to
         
        <a
          class="sc-dlniIP LWZKO"
          href="https://nextjs.org"
        >
          Next.js!
        </a>
      </h1>
      <p
        class="sc-jrsKJM eHsNYH"
      >
        Get started by editing
        <code
          class="sc-fujznN SvhJn"
        >
          pages/index.tsx
        </code>
      </p>
      <div
        class="sc-gKAaef cJkFME"
        spaces="2"
      />
      <div>
        <input
          data-testid="checkbox-acceptTerms"
          id="acceptTerms"
          type="checkbox"
        />
        <label
          for="acceptTerms"
        >
          Checkbox à cocher
        </label>
      </div>
      <div
        class="sc-hKFymg jLSsZz"
      >
        <div
          class="sc-eCAqax blQmze"
        >
          <a
            class="sc-jSFipO fkzsOZ"
            href="/about"
          >
            About →
          </a>
        </div>
      </div>
      <div
        class="sc-gKAaef cJkFME"
        spaces="2"
      />
      <button>
        Trigger analytics event
      </button>
    </main>
  </div>
</div>
`;
